---
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: fill
    social: menu
    theme: united
params:
  set_title: scATAC-pro Report
  set_sample: PMBC10K
  output_dir: /mnt/isilon/tan_lab/yuw1/run_scATAC-pro/PBMC10k/output
  configure_user: /mnt/isilon/tan_lab/yuw1/run_scATAC-pro/PBMC10k/configure_user.txt
title: "scATAC-pro `r params$set_title`: `r params$set_sample`"
---

<style type="text/css">

.title { /* Header 1 */
  font-size: 28px;
}

.chart-title {  /* chart_title  */
   font-family: "Raleway", Helvetica, Arial, sans-serif;
   font-size: 18px;
   font-weight: 800;
}  
.navbar-brand {
   font-size: 22px;
   font-weight: 900;
}

</style>


<style type="image-container">
  width: 100%;
  margin-left: 0;
  margin-right: 0;
</style>



```{r setup, include=FALSE}
library(flexdashboard)
library(data.table)
library(magrittr)
library(kableExtra)
library(RColorBrewer)
library(ggplot2)
library(cicero)
library(viridis)
library(GenomicRanges)

```

<span style="font-size:20px;"> Global QC </span>
==========================================

Column
-----------------------------------------------------------------------

### Sequencing: Overall

```{r}

read_conf_file <- function(configure_user_file){
   temp1 <- tempfile()
   temp2 <- tempfile()
   pr1 = paste0(" | grep -v ^# | awk -F= '{print $1}' | awk '{$1=$1;print}' >", temp1)
   pr2 = paste0(" | grep -v ^# | awk -F= '{print $2}' | awk -F# '{print $1}' | awk '{$1=$1;print}' >", temp2)
  system(paste0('grep = ', configure_user_file, pr1))
  system(paste0('grep = ', configure_user_file, pr2))


  vrs = readLines(temp1)
  vls = readLines(temp2)
  unlink(temp1)
  unlink(temp2)
  for(i in 1:length(vrs)){
    assign(vrs[i], vls[i], envir = .GlobalEnv)
  }
}

dir.create(paste0(params$output_dir, '/summary/Tables'), showWarnings = F)

read_conf_file(params$configure_user)
plotEPS = as.logical(plotEPS)
if(plotEPS) dir.create(paste0(params$output_dir, '/summary/Figures'), showWarn = F)
mapping_qc_file = paste0(params$output_dir, '/summary/', OUTPUT_PREFIX,  '.MappingStats')

fragments_file = paste0(params$output_dir, '/summary/', OUTPUT_PREFIX,  '.fragments.tsv.gz')
fragments_len_file = paste0(params$output_dir, '/summary/', OUTPUT_PREFIX,  '.fragments.len.tsv.gz')
frags = fread(fragments_file)
names(frags) = c('chr', 'start', 'end', 'bc', 'ndup')


mapping_qc = fread(mapping_qc_file, header = F)

lib_complx = nrow(frags)/sum(frags$ndup)
lib_complx = paste0(100 * round(lib_complx, 3), '%')

mapping_qc$frac = round(mapping_qc$V2/mapping_qc$V2[1], 3)
mapping_qc$frac = paste0(100*mapping_qc$frac, '%')

mapping_qc = rbind(mapping_qc, data.frame(V1 ='Library Complexity (#unique fragments/#fragments)', V2 = '', frac = lib_complx))

kable(mapping_qc, col.names = NULL, format = 'html') %>%
  kable_styling("striped", full_width = F, position = 'left', font_size = 15)

write.table(mapping_qc, file = paste0(params$output_dir, '/summary/Tables/Global_Mapping_Statistics.tsv'),
	    col.names = F, row.names = F, sep = '\t', quote = F)
```


Column
-----------------------------------------------------------------------


### Cell/Peak Calling Summary 

```{r}
cell_mapping_qc_file = paste0(params$output_dir, '/summary/cell_barcodes.MappingStats')

if(file.exists(cell_mapping_qc_file)) cell_mapping_qc = fread(cell_mapping_qc_file, header = F)

bc_stat_file = paste0(params$output_dir, '/summary/', OUTPUT_PREFIX, '.', PEAK_CALLER, '.qc_per_barcode.txt')
selected_bcs = paste0(params$output_dir, '/filtered_matrix/', PEAK_CALLER, '/', CELL_CALLER, '/barcodes_doubletsRemoved.txt')
if(!file.exists(selected_bcs)) selected_bcs = paste0(params$output_dir, '/filtered_matrix/', PEAK_CALLER, '/', CELL_CALLER, '/barcodes.txt')

peak_file = paste0(params$output_dir, '/peaks/', PEAK_CALLER, '/', OUTPUT_PREFIX, '_features_BlacklistRemoved.bed')

bc_stat = fread(bc_stat_file)

barcodes = fread(selected_bcs, header = F)$V1

peaks = fread(peak_file)

qc_sele = bc_stat[bc %in% barcodes, ]
qc_nonsele = bc_stat[!bc %in% barcodes, ]


#frags = frags[bc %in% barcodes]

ncells = length(barcodes)
mapq30.frac.in.cell = paste0(round(sum(frags[bc %in% barcodes]$ndup)/sum(frags$ndup), 3) * 100, '%')

#frac.in.cell = paste0(round(cell_mapping_qc$V2[2]/as.integer(as.character(mapping_qc$V2[2])), 3) * 100, '%')

med.frag.per.cell = round(median(qc_sele$total_frags))
frac.uniq = paste0(round(nrow(frags[bc %in% barcodes])/sum(frags[bc %in% barcodes]$ndup), 3) * 100, '%')

#frac.in.peak = sum(bc_stat$total_frags * bc_stat$frac_peak)/sum(bc_stat$total_frags)
names(peaks)[1:3] = c('chr', 'start', 'end')
frac.in.peak <- GenomicRanges::countOverlaps(makeGRangesFromDataFrame(frags), makeGRangesFromDataFrame(peaks))
frac.in.peak <- mean(frac.in.peak > 0)
frac.in.peak = paste0(round(frac.in.peak, 3)*100, '%')

#frac.in.cell = sum(qc_sele$total_frags)/sum(bc_stat$total_frags)
frac.in.cell = sum(qc_sele$total_frags)/nrow(frags)
frac.in.cell = paste0(round(frac.in.cell, 3)*100, '%')

cell.table = data.frame(c(PEAK_CALLER, nrow(peaks),
                          frac.in.peak, 
                          CELL_CALLER,  paste0(ncells),
                          frac.in.cell,
                          paste0(med.frag.per.cell), 
                          frac.in.cell, mapq30.frac.in.cell)) 

rm(peaks)

rownames(cell.table) = c('Peak called by', 'Number of peaks', 
                         'Percentage of unique fragments in peaks',
                         'Cell called by', 'Estimated # of cells',
                         'Percentage of unique fragments in cells',
                         'Median # of unique fragments per cell', 
                         'Percentage of Mapped fragments in cells', 
                         'Percentage of MAPQ30 fragments in cells')  

kable(cell.table, row.names = T, col.names = NULL, format = 'html') %>%
  kable_styling("striped", full_width = F, position = 'left', font_size = 15)

write.table(cell.table, file = paste0(params$output_dir, '/summary/Tables/CellAndPeak_Calling_Summary.tsv'),
	   col.names = F, quote = F, sep = '\t')

```

Column
-----------------------------------------------------------------------

### Sequencing: reads from cell barcodes

```{r}
if(as.logical(CELL_MAP_QC) & file.exists(cell_mapping_qc_file)){
    lib_complx = frac.uniq

    cell_mapping_qc$frac = round(cell_mapping_qc$V2/cell_mapping_qc$V2[1], 3)
    cell_mapping_qc$frac = paste0(100*cell_mapping_qc$frac, '%')

    cell_mapping_qc = rbind(cell_mapping_qc, data.frame(V1 ='Library Complexity (#unique fragments/#fragments)', V2 = '', frac = lib_complx))

    write.table(cell_mapping_qc, file = paste0(params$output_dir, '/summary/Tables/Cell_Mapping_QC.tsv'),
		col.names = F, quote = F, row.names = F, sep = '\t')

    kable(cell_mapping_qc, row.names = F, col.names = NULL, format = 'html') %>%
      kable_styling("striped", full_width = F, position = 'left', font_size = 15)


}else{
    cat('Sequencing Alignment QC for reads in cell barcodes was not generated!')
    cat('To generate it, 1) run module get_bam4Cells, 2) run module report again')
}


```



<span style="font-size:20px;"> Cell Barcode QC  </span> {data-orientation=rows}
==========================================

Row 
--------------------------------------------------------------------


### Total fragments VS fraction in peaks

```{r}

bc_stat[, 'group' := ifelse(bc %in% barcodes, 'cell', 'non-cell')]

library(ggplot2)
library(grid)
nsub_frags = min(15000, nrow(bc_stat))
cellColors = c('cell' = '#F8766D', 'non-cell' = '#00BFC4')
g <- ggplot(data = bc_stat[sample(1:nrow(bc_stat), nsub_frags), ], 
            aes(x = total_frags, y = frac_peak, col = group)) +
     scale_colour_manual(values = cellColors) +  
  geom_point(size = 0.5) + scale_x_continuous(trans='log10') + theme_bw() +
      theme(legend.position = 'none', 
            legend.title=element_blank(),
            axis.text = element_text(size = 15, family = "Helvetica"),
            axis.title = element_text(size = 18, family = "Helvetica")) +
  xlab('Total #Unique Fragments') + ylab('Fraction in Peak') 


text1 <- grobTree(textGrob("Cell", x=0.8,  y=0.93, hjust=0,
  gp=gpar(col='#F8766D', fontsize=15, fontface = 'bold', fontfamily = "Helvetica")))
text2 <- grobTree(textGrob("Non-cell", x=0.8,  y=0.83, hjust=0,
  gp=gpar(col='#00BFC4', fontsize=15, fontface = 'bold', fontfamily = "Helvetica")))

g <- g + annotation_custom(text1) + annotation_custom(text2)
g



```



### Distribution of Insert Size (All Barcodes) 

```{r}
if(file.exists(fragments_len_file)) {
    frag_lens = fread(fragments_len_file)
    names(frag_lens) = 'isize'
}else{
    lens = frags$end - frags$start
    frag_lens = data.table('isize' = rep(lens, frags$ndup))
}
rm(frags)
frag_lens = frag_lens[sample(1:nrow(frag_lens), 1000000), ]

p1 <- ggplot(data = frag_lens[isize < 800], aes(x = isize)) +
      geom_density(fill = 'lightblue') + xlab('Insert Size') + ylab('Density') + theme_bw() + 
      theme(legend.title=element_blank(), 
                    legend.background = NULL, 
                    axis.text = element_text(size = 15, family = "Helvetica"), 
                    axis.title = element_text(size = 18, family = "Helvetica")) 

p1

if(plotEPS){
  pfname = paste0(params$output_dir, '/summary/Figures/dist_insert_size.eps')
  ggsave(p1, file = pfname, device = 'eps', width = 6, height = 6)
}



```


### Density plot of total number of unique fragments

```{r}

bc_stat[, 'group' := ifelse(bc %in% barcodes, 'cell', 'non-cell')]

p <- ggplot(data = bc_stat, aes(x = total_frags, fill = group)) + 
  geom_density() + scale_x_continuous(trans = 'log10') + theme_bw() +
  theme(legend.position='none', legend.title=element_blank(),
        axis.title = element_text(size = 18, family = "Helvetica"),
        axis.text = element_text(size = 15, family = "Helvetica")) + 
  xlab('Total #Unique Fragments') + ylab('Density') 

text1 <- grobTree(textGrob("Cell", x=0.8,  y=0.93, hjust=0,
  gp=gpar(col='#F8766D', fontsize=15, fontface = 'bold', fontfamily = "Helvetica")))
text2 <- grobTree(textGrob("Non-cell", x=0.8,  y=0.83, hjust=0,
  gp=gpar(col='#00BFC4', fontsize=15, fontface = 'bold', fontfamily = "Helvetica")))

p <- p + annotation_custom(text1) + annotation_custom(text2)

p

if(plotEPS){
  system(paste0('mkdir -p ', params$output_dir, '/summary/Figures'))
  pfname = paste0(params$output_dir, '/summary/Figures/dist_frags.eps')
  ggsave(p, file = pfname, device = 'eps', width = 6, height = 6)
}


```




Row 
-------------------------------------------------------------------

### TSS enrichment score profile (Cell Barcodes)

```{r}

tss_escore_file = paste0(params$output_dir, '/signal/cell_barcodes.MAPQ30.aggregated.mtx.gz')
if(!file.exists(tss_escore_file)) tss_escore_file = paste0(params$output_dir, '/signal/', OUTPUT_PREFIX, '.aggregated.mtx.gz')

if(file.exists(tss_escore_file)){
    set.cols = brewer.pal(n=5, name = 'Dark2')
    tss.mat = fread(tss_escore_file)
    tss.mat = tss.mat[, -c(1:6)]
    tss.mat[is.na(tss.mat)] = 0
    tss.escore = colSums(tss.mat)
    ma <- function(x, n = 10){stats::filter(x, rep(1 / n, n), sides = 2)}
    tss.escore = ma(tss.escore)
    tss.escore = tss.escore[14:213]
    df = data.table(index = 10*(-100:99), escore = tss.escore/tss.escore[1])

    p0 <- ggplot(data = df, aes(x = index, y = escore)) + geom_line(size = 1, col = set.cols[1])  + theme_bw() +
      xlab('Distance to TSS (bp)') + ylab('TSS enrichment score') + theme(legend.title=element_blank(),
          axis.text = element_text(size = 15, family = "Helvetica"), 
          axis.title = element_text(size = 18, family = "Helvetica")) 

    print(p0)


    if(plotEPS){

      pfname = paste0(params$output_dir, '/summary/Figures/tss_enrich.eps')
      ggsave(p0, file = pfname, device = 'eps', width = 6, height = 6)
    }
}

```



### Overlapping with sequence annotated regions (Cell Barcodes)
```{r}

qc_sele_df = data.table(frac = c(qc_sele$frac_peak, qc_sele$frac_tss, qc_sele$frac_promoter, qc_sele$frac_enh, qc_sele$frac_mito), 'type' = rep(c('Peaks', 'Tss', 'Promoter', 'Enhancer', 'Mito'), each = nrow(qc_sele)))

qc_sele_df$type = factor(qc_sele_df$type, levels = c('Peaks', 'Tss', 'Promoter', 'Enhancer', 'Mito'))

p0 <- ggplot(data = qc_sele_df, aes(y = frac, x = type, fill = type)) + ylab('Fraction') + theme_bw() +
  geom_boxplot(outlier.size = 0.01, show.legend = FALSE) +  
  theme(legend.position = 'none', 
        axis.text = element_text(size = 18, family = "Helvetica"), 
        axis.title.x = element_blank(), 
        axis.title.y = element_text(size = 18, family = "Helvetica")) + xlab('') 
p0

if(plotEPS){
  pfname = paste0(params$output_dir, '/summary/Figures/overlap_annotation.eps')
  ggsave(p0, file = pfname, device = 'eps', width = 6, height = 6)
}


```

### Overall statistics (Cell Barcodes)

```{r}

frac_peak = sum(qc_sele$total_frags * qc_sele$frac_peak)/sum(qc_sele$total_frags)
frac_mito = sum(qc_sele$total_frags * qc_sele$frac_mito)/sum(qc_sele$total_frags)
frac_promoter = sum(qc_sele$total_frags * qc_sele$frac_promoter)/sum(qc_sele$total_frags)
frac_enh = sum(qc_sele$total_frags * qc_sele$frac_enhancer)/sum(qc_sele$total_frags)
frac_tss = sum(qc_sele$total_frags * qc_sele$frac_tss)/sum(qc_sele$total_frags)

frac_peak = median(qc_sele$frac_peak)
frac_mito = median(qc_sele$frac_mito)
frac_promoter = median(qc_sele$frac_promoter)
frac_enh = median(qc_sele$frac_enhancer)
frac_tss = median(qc_sele$frac_tss)
fracs = data.frame(c(frac_peak,  frac_promoter, frac_enh, frac_tss, frac_mito))
rownames(fracs) = c('Median % fragments overlapping peaks', 
                    'Median % fragments overlapping promoters', 
                    'Median % fragments overlapping Enhancers(ENCODE)', 
                    'Median % fragments overlapping TSS (+/-1Kb)', 
                    'Median % fragments in mitochondrial genome')
colnames(fracs) = 'pr'
fracs$pr = round(fracs$pr, 3)
fracs$pr = paste0(100*fracs$pr, '%')

kable(fracs, row.names = T, col.names = NULL) %>%
  kable_styling(full_width = F, position = 'left', font_size = 15)


write.table(fracs, file = paste0(params$output_dir, '/summary/Tables/Cell_FunctionRegion_Statistics.tsv'),
		col.names = F, quote = F, sep = '\t')

```



<span style="font-size:20px;"> Downstream Analysis </span>
=====================================================


Column
-----------------------------------------------------

### Clustering 
```{r}
library(Seurat)

down.dir = paste0(params$output_dir, '/downstream_analysis/', PEAK_CALLER, '/', CELL_CALLER)
seurat_file = paste0(down.dir, '/seurat_obj.rds')
if(file.exists(seurat_file)){
    ss = readRDS(seurat_file)

    cg <- DimPlot(ss, reduction = 'umap', group.by = 'active_clusters', label = T) + scale_color_brewer(palette = "Paired") + theme(legend.text = element_text(size = 18, family = "Helvetica"))
    
    
    
    if(plotEPS){
      pfname = paste0(params$output_dir, '/summary/Figures/umap_clusters.eps')
      ggsave(cg, file = pfname, device = 'eps', width = 6, height = 6)
    }
    
    cg
}


```


### GO Analysis
```{r, warning=F, message=F}
go_file = paste0(down.dir, '/enrichedGO_differential_accessible_features_', group1, 
                  '_vs_', group2, '.xlsx')
if(file.exists(go_file)){
    if(grepl(group1, pattern = 'one')){
      tmp = list()
      wb <- xlsx::loadWorkbook(go_file)
      sheets = xlsx::getSheets(wb)
      for(i in names(sheets)) tmp[[i]] = xlsx::read.xlsx(go_file, sheetName = i)
      ll = sapply(tmp, nrow)
      go_res = tmp[[which.max(ll)]]
      group.sele = names(sheets)[which.max(ll)]
      group.sele = gsub('cluster', '', group.sele)
    }else{
       group1 = gsub(':', '_', group1)
       group2 = gsub(':', '_', group2)
       go_res1 = xlsx::read.xlsx(go_file, sheetName = paste0('cluster', group1))
       go_res2 = xlsx::read.xlsx(go_file, sheetName = paste0('cluster', group2))
       go_res = go_res1
       group.sele = group1
       if(nrow(go_res1) < nrow(go_res2) & group2 != 'rest'){
          go_res = go_res2
          group.sele = group2
       }
    }
    
    
    if(nrow(go_res) == 0) {
        if(grepl(group1, pattern = 'one')) {
            cat(paste0('No enriched terms in comparison of ', group1, '_vs_', group2))
        }else{
            cat(paste0('No enriched terms in comparison of cluster_', group1, '_vs_', group2))
        }
    }else{
        #go_res = fread(go_file)
        go_res = data.table(go_res)
        go_res[, 'score' := -log10(p.adjust)]
        go_res = go_res[order(-score), ]
        ngo = min(20, nrow(go_res))
        go_res = go_res[1:ngo, ]
        go_res = go_res[order(score), ]
        go_res$Description = as.character(go_res$Description)
        go_res$Description = sapply(go_res$Description, function(x) substr(x, 1, min(50, nchar(x)))) 
        go_res$Description = factor(go_res$Description, levels = go_res$Description)
        
        p_go <- ggplot(go_res, aes(y = score, x = Description, fill = Count)) +
          geom_bar(width = 0.7, stat = 'identity') +
          ggtitle(paste0("Enriched terms: cluster_", group.sele)) + theme_classic() + 
          theme(legend.position = 'bottom', legend.direction = "horizontal") + 
          coord_flip()  +  scale_fill_continuous(name = "#genes", type = "viridis") +
          xlab('') + ylab('-log10(p.adjust)')
         
        print(p_go) 
        
        if(plotEPS){
            pfname = paste0(params$output_dir, '/summary/Figures/enriched_GO.eps')
            ggsave(p_go, file = pfname, device = 'eps', width = 10, height = 6)
        }
    }
}else{
    cat("No Go Enrichment result for given comparison, please run the runGO module of scATAC-pro first!")
}

```


Column
----------------------------------------------------

### Motif Enrichment Analysis

```{r, fig.height=9, out.width='\\textwidth'}
## check enriched TFs for each cluster
library(chromVAR)
library(BiocParallel)
register(SerialParam())

# Do DA/DE with one cluster vs the rest clusters
# clusters are the data frame with <barcode> <cluster>
do_DA_motif <- function(mtx_score, clusters, test = 'wilcox', 
                  only.pos = T, fdr = 0.05, topn = 10){
  clusters$cluster = as.character(clusters$cluster)
  cls = unique(clusters$cluster)
  res = NULL
  features = rownames(mtx_score)
  for(cluster0 in cls){
    bc0 = clusters[cluster == cluster0]$barcode
    mtx1 = mtx_score[, colnames(mtx_score) %in% bc0]
    mtx2 = mtx_score[, !colnames(mtx_score) %in% bc0]
    mu1 = sapply(1:length(features), function(x) mean(mtx1[x, ]))
    mu2 = sapply(1:length(features), function(x) mean(mtx2[x, ]))
    
    pvs = rep(0.5, length(features))
    
    for(x in 1:length(features)){
      a1 = mtx1[x, ]
      a2 = mtx2[x, ]
      if(length(which(!is.na(a1))) < 2 || length(which(!is.na(a2))) < 2) next
      pvs[x] = wilcox.test(a1, a2, alternative = 'greater')$p.value
    }
    
    pvs.adj = p.adjust(pvs, method = 'fdr')
    res0 = data.table('feature' = features, 'cluster' = cluster0,
                      'mean1' = mu1, 'mean2' = mu2,
                       'pv' = pvs, 'pv_adjust' = pvs.adj)
    
    
    res0 = res0[order(pv_adjust), ]
    res0 = res0[pv_adjust <= fdr]
    
    if(nrow(res0) > topn) res0 = res0[1:topn, ]
    res = rbind(res, res0)
  }
  return(res)
}


if(file.exists(seurat_file)){
    metaData = ss@meta.data
    rm(ss)
}

if(file.exists(paste0(down.dir, '/chromVar_obj.rds'))){
  chromVar.obj = readRDS(paste0(down.dir, '/chromVar_obj.rds'))
  if(!file.exists(paste0(down.dir, '/differential_TF_motif_enriched_in_clusters.tsv'))){
         dev = deviations(chromVar.obj)
      da.res = do_DA_motif(dev, 
                 clusters = data.table('barcode' = rownames(metaData),
                                       'cluster' = metaData$active_clusters),
                 topn = 10)
      rm(dev)
      write.table(da.res, file = paste0(down.dir, '/differential_TF_cluster_enrich.tsv'), 
                  sep = '\t', quote = F, row.names = F )
  }else{
    da.res = fread( paste0(down.dir, '/differential_TF_motif_enriched_in_clusters.tsv'))
  }
  
  ## plot enriched TFs in heatmap
  sele.tfs = da.res$feature
  #zscores = chromVar.obj@assays$data$z
  zscores = deviationScores(chromVar.obj)
  sele.zscores = zscores[sele.tfs, ]
  
  
  # change tf name to be more readable
  if(grepl(GENOME_NAME, pattern = 'hg', ignore.case = T)){
    rnames = rownames(sele.zscores)
    nnames = sapply(rnames, function(x) unlist(strsplit(x, '_'))[3])
    nnames1 = sapply(rnames, function(x) unlist(strsplit(x, '_'))[1])
    rownames(sele.zscores) = ifelse(grepl(nnames, pattern = 'LINE'), nnames1, nnames)
  }else{
    rnames = rownames(sele.zscores)
    nnames = sapply(rnames, function(x) unlist(strsplit(x, '_'))[3])
    rownames(sele.zscores) = nnames
    sele.zscores = sele.zscores[!grepl(nnames, pattern = '^LINE'), ]
  }
  
  
  metaData$active_clusters = as.character(metaData$active_clusters)
  metaData = data.table(metaData, keep.rownames = T)
  setkey(metaData, active_clusters)
  
  rr = metaData$rn[metaData$rn %in% colnames(sele.zscores)]
  sele.zscores = sele.zscores[, rr]
  
  
  sele.zscores = sele.zscores[!duplicated(sele.zscores), ]
  
  ann_col = data.frame('cluster' = metaData$active_clusters)
  rownames(ann_col) = metaData$rn
  
  up_cut = quantile(sele.zscores, 0.95, na.rm = T)
  low_cut = quantile(sele.zscores, 0.05, na.rm = T)
  sele.zscores[is.na(sele.zscores)] = 0
  low_cut = min(0, low_cut)
  sele.zscores[sele.zscores > up_cut] = up_cut
  sele.zscores[sele.zscores < low_cut] = low_cut
  
  cluster = brewer.pal(n=length(unique(metaData$active_clusters)), name = 'Paired')
  names(cluster) = sort(unique(metaData$active_clusters))
  ann_colors = list('cluster' = cluster)
  
  # resample to reduce memory used
  set.seed(2019)
  rids = sort(sample(1:ncol(sele.zscores), floor(ncol(sele.zscores)/6)))
  ann_col0 = data.frame(ann_col[rids, ])
  rownames(ann_col0) = colnames(sele.zscores)[rids]
  mtx0 = sele.zscores[, rids]
  names(ann_col0) = 'cluster'
  ph <- pheatmap::pheatmap(mtx0, cluster_cols = F,
                     cluster_rows = T, show_colnames = F, fontsize = 13,
                     annotation_col = ann_col0, color = viridis(100),
                     annotation_colors = ann_colors, fontsize_row = 9)
  
  if(plotEPS){
  pfname = paste0(params$output_dir, '/summary/Figures/heatmap_motif_enrich.eps')
  #postscript(file = pfname, width = 9, height = 12)
  
  ggsave(ph, filename = pfname, device = 'eps', height = 12,
         width = 9)
  #dev.off()
  }

}


```

Column
----------------------------------------------------


### Footprinting Analysis: differentially bound TFs

```{r, out.width='\\textwidth'}

footprint_stats.file = paste0(down.dir, '/differential_TF_footprint_', 
                              group1_fp, '_vs_', group2_fp, '.tsv')
if(file.exists(footprint_stats.file)){
  
  footprint_out = fread(footprint_stats.file)
  if(length(unique(footprint_out$motif)) > 100){
  footprint_out[, 'N' := .N, by = higher_in_cluster]
  cls = unique(footprint_out[N > 10]$higher_in_cluster)
  if(length(cls) >= 1){
    res0 = NULL
    for(cl0 in cls){
      tmp = footprint_out[higher_in_cluster == cl0]
      tmp = tmp[order(P_values)][1:10, ]
      res0 = rbind(res0, tmp)
    }
    footprint_out = rbind(footprint_out[N < 10], res0)
  }
}

  mm = reshape2::acast(motif ~ higher_in_cluster, data = footprint_out, 
                       value.var = "P_values")
  mm = -log10(mm)
  mm[is.na(mm)] = 0
  cn = colnames(mm)
  cn.new = sapply(cn, function(x) gsub('_higher', '', x))
  colnames(mm) = cn.new
  mm[mm > 3] = 3
  p1 <- pheatmap::pheatmap(mm, cluster_cols = F, fontsize = 13, fontsize_row = 9,
                           color = viridis::viridis(100))
  
  pfname1 =paste0(params$output_dir, '/summary/Figures/heatmap_differential_TF_footprint_', 
                  group1_fp, '_vs_', group2_fp, '.eps')
  
  if(plotEPS){
    ggsave(p1, filename = pfname1, device = 'eps', height = 6,
           width = 6)
  }
  
  p1

}



```


### Predicted Interactions at a locus of interest
```{r, out.width='\\textwidth'}
cicero_conn.file = paste0(down.dir, '/cicero_interactions.txt')
if(Cicero_Plot_Region != 'none' & file.exists(cicero_conn.file)){
 conns = fread(cicero_conn.file)
 temp <- tempfile()
  if(grepl(GENOME_NAME, pattern = 'mm10', ignore.case = T)) {
    library(EnsDb.Mmusculus.v79)
    ens.ann <- EnsDb.Mmusculus.v79
  }
  if(grepl(GENOME_NAME, pattern = 'mm9', ignore.case = T)) {
    library(EnsDb.Mmusculus.v75)
    ens.ann <- EnsDb.Mmusculus.v75
  }
  
  if(grepl(GENOME_NAME, pattern = 'hg38', ignore.case = T)) {
    library(EnsDb.Hsapiens.v86)
    ens.ann <- EnsDb.Hsapiens.v86
  }
  
  if(grepl(GENOME_NAME, pattern = 'hg19', ignore.case = T)) {
    library(EnsDb.Hsapiens.v75)
    ens.ann <- EnsDb.Hsapiens.v75
  }
  
  gene_anno <- transcripts(ens.ann,
                          columns = c("gene_id", "gene_name",  
                                      "tx_biotype", "tx_id"))
 
  gene_anno = keepStandardChromosomes(gene_anno,pruning.mode = 'coarse')
 
  gene_anno = data.table(data.frame(gene_anno))
  gene_anno = subset(gene_anno, tx_biotype %in% c('protein_coding', 'miRNA', 'lincRNA'), select = c('seqnames', 'start', 'end', 'tx_id',
                       'strand', 'gene_id', 'gene_name'))
  #rename some columns to match requirements
  names(gene_anno)[c(1, 4, 6:7)] = c('chromosome', 'transcript', 'gene', 'symbol')
  gene_anno$chromosome = paste0('chr', gene_anno$chromosome)

  gene_anno = subset(gene_anno, select = c(chromosome, start, end, strand, 
                                           transcript, gene, symbol))
  gene_anno = gene_anno[complete.cases(gene_anno), ]
  if(Cicero_Plot_Region %in% gene_anno$symbol){
    tmp_anno = gene_anno[gene_anno$symbol == Cicero_Plot_Region, ]
    chr0 = tmp_anno$chr[1]
    start0 = min(tmp_anno$start)
    end0 = max(tmp_anno$end)
  }else{
    chr0 = unlist(strsplit(Cicero_Plot_Region, ':'))[1] ## chr5:140610000-140640000
    region0 = unlist(strsplit(Cicero_Plot_Region, ':'))[2]
    start0 = as.integer(unlist(strsplit(region0, '-'))[1])
    end0 = as.integer(unlist(strsplit(region0, '-'))[2])
  }
  
  if(end0 < start0 + 100000){
    start0 = floor((start0+end0)/2) - 50000
    end0 = start0 + 100000
  }
  conns[, 'chr1' := unlist(strsplit(Peak1, '_'))[1], by = Peak1] 
  conns[, 'start1' := as.integer(unlist(strsplit(Peak1, '_'))[2]), by = Peak1] 
  conns[, 'end2' := as.integer(unlist(strsplit(Peak2, '_'))[3]), by = Peak2] 
  conns = conns[start1 <= end2]
  conns0 = conns[chr1==chr0 & start1 >= start0 & end2 <= end0]
  
  conns[, c('chr1', 'start1', 'end2') := NULL]
  conns = data.frame(conns)
  if(nrow(conns0)==0) cat('No interaction found in the griven region!')
  
  if(plotEPS & nrow(conns0) >0 ){
    pdf(paste0(params$output_dir,'/summary/Figures/interactions_example_region.pdf'), width = 8, height = 5, fonts = 'sans')
    plot_connections(conns, chr0, start0, end0,
                     gene_model = gene_anno, 
                     coaccess_cutoff = .3, 
                     connection_width = 1, 
                     collapseTranscripts = "longest",
                     viewpoint_alpha = 0)
    dev.off()
    
  }
  if(nrow(conns0) > 0)  plot_connections(conns, chr0, start0, end0,
                                         gene_model = gene_anno, 
                                         coaccess_cutoff = .3, 
                                         connection_width = 1, 
                                         collapseTranscripts = "longest",
                                         viewpoint_alpha = 0)
}


```
